### ISSUE TITLE
Regarding deltabot not ignoring tokens within quote or code markup blocks

### ISSUE THREAD

**Issue by s33plusplus** [MystK/delta-bot-three#3]
Hey MystK, I've been working with Snorrrlax and RustyRook over on r/CMV to fix a comment parsing bug in Deltabot 2.0, and Snorrrlax has asked me to report what I found as well as a summary of the fix to avoid the issue in the future.

So, after digging into MARS last night, I tracked down the function that runs a check against the comment body looking for code and quote markup with a simple regex (search_line() in modules/comments.py). Backtracking to where this function is called, I noticed it was passing the body attribute of the comment object (praw.objects.Comment) and spotted the problem. 

In the current version of that search_line function, the regex used to match the markup is written like so:
`"(    |&gt;)"`

The issue is the input for that function doesn't contain the semi-HTML escaped version of the comment body, but the raw comment body as it was typed into the reply box on reddit, which means `"&gt;"` will never actually match a code block. Additionally, due to the way reddit interprets the markup, any code or quote block will always be on a single line with the delimiter right at the start of the line. 

Here's the regex I tweaked a bit to get search_line matching those delimiters correctly:
`"(^    |^>)"`

I don't actually have access to a test subreddit to run the bot with the fix in place, so I've only simulated this pair of functions by calling them with PRAW objects I retrieved myself, in more or less the same way the bot does with a couple handpicked comments to test against. You can find the patch I wrote over at my [fork of MARS](https://github.com/s33plusplus/MARS).

I think the entire issue here came from confusion about what type of data the reddit API passes back when requesting comments. From my experience with both PRAW and the raw JSON, there are always two representations of the comment body; the verbatim copy of what the user typed into the reply box (referred to as "body"), and a HTML copy rendered from the aforementioned raw comment (referred to as "body_html"). 

I've found it much easier to just outright avoid the HTML version if at all possible since it simplifies everything, and when you do have to touch the HTML version, keep the code that handles it isolated and clearly marked. There are all sorts of inconsistencies in the reddit API and the data it returns, I've learned the hard way that you've gotta keep it simple and test the hell out of the code to catch the wild bugs that pop up.

Anyway, thanks for all the hard work you and the rest of the mod team put into r/CMV! If you need to reach me over on reddit, I use the same handle (u/s33plusplus) because naming things is hard.

**Comment by MystK** [comment]
Hey man! Thanks for all the good info here. From a quick skim, I believe I might have already encountered and fixed this in #1. I'll leave this open for you until I have time to fully analyze this and make sure I've covered everything you mentioned. Thanks! :)

**Comment by s33plusplus** [comment]
No problem, happy to help! So, I'm going to be upfront, I don't have much experience with JS, so it's taking me a bit longer to read through the code and I may make some mistakes, so I apologise in advance.

Looking through that issue, it seems like it's covering one particular case involving the unicode symbol, but does the regex also catch other quoted tokens? As an example of what I mean, if I were to put the following comment into a thread:

`> !delta`

Would that be safe to use for demonstration purposes? The issue as explained to me by RustyRook over on the subreddit was deltabot is supposed to ignore quoted or code-blocked tokens to allow users to show other users how to trigger deltabot without accidentally doing it themselves, but due to a bug deltabot will still trigger on any mention of the "!delta" token. [Here's the exchange that prompted the bug hunt](https://www.reddit.com/r/changemyview/comments/4c29fg/cmv_white_people_can_be_victims_of_racism/d1elvm4) for the full context. 

Thanks for the quick followup, and seriously, mad props for rewriting such a busy bot from scratch like this!

**Comment by MystK** [comment]
Hey! Thanks for the example! DB3 currently doesn't take into account deltas in quotes! That's a great example of a use case. I'd have to edit the regular expression to ignore deltas between the '>' and new lines. It shouldn't be too big of an issue.

Date: Mon, 28 Mar 2016 12:52:44 -0700

**Comment by s33plusplus** [comment]
Awesome, solid plan! For the sake of completeness, there is one other character sequence that someone would use similarly to the '>' character in a post; the code block. Similar to the previous example, this should ideally also be ignored (c/python style string escaped to avoid whitespace filtering):

`"\s\s\s\s!delta\n"`

In case that gets mangled, it's 4 single space characters at the start of the line, and runs until it hits a newline just like the previous quote example. With the equivalent of `(^ {4}|^>).+\n$` (excuse my regex mistakes, still learning) in there, it should cover both of the ways a user would try to keep deltabot from interpreting a token unintentionally, as my patch to deltabot 2.0/MARS was attempting to accomplish.

**Comment by MystK** [comment]
Thanks! I'll be sure to ignore deltas in quotes and code blocks. I will close this ticket once it is complete

**Comment by MystK** [comment]
Hi @s33plusplus, I did some testing and it looks like the current code already ignores code blocks and quotes. I did not add any extra code to fix this, but I may have already done so in the beginning (I forgot).

I will be closing this. If someone tests  a corner and it is still happening, this ticket will be reopened or a new one will be made.
